[1-3단계 - 방탈출 사용자 예약] 줄리(한승아) 미션 제출합니다.

jbilee opened the pull request:
## 수정 코드 범위
이전 미션에서 불러온 코드를 추가한 커밋을 제외한 변경 범위입니다:
https://github.com/woowacourse/spring-roomescape-member/pull/244/files/0361ea69a2b23fec2ce8b5d3d9686e9b8d74face..7836b55630be7a303a939cf0fd82a1b7a0f89f5b

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  https://prolog.techcourse.co.kr/studylogs/4278

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
메서드 분리가 필요한 부분이 남아있어서 선택했습니다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [x] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
👋 안녕하세요, 솔라!

이번에 방탈출 사용자 예약 미션에 대한 리뷰를 요청하게 된 줄리입니다. ☺️
미션을 진행하면서 아래와 같은 고민을 하게 되었습니다.

### 엔드포인트 공유
구현한 컨트롤러 중 테마 API와 예약 시간 API의 컨트롤러에는 한 엔드포인트에 각각 다른 HTTP 메서드가 매핑되어 있습니다. 예를 들어서 `themes/popular`에 GET 요청을 보내면 인기 테마 목록을 반환하는 서비스 메서드를 호출하고, `themes/popular`에 DELETE 요청을 보내면 popular라는 id를 가진 리소스를 삭제하는 서비스 메서드를 호출하는 구조입니다.

올바른 메서드로 요청을 보내면 문제될 것이 없지만, 한 엔드포인트가 두가지 일을 하는 것으로 보여 고민하게 되었습니다. 그렇다고 컨트롤러를 분리하면 동일한 theme 모델에 대한 컨트롤러인데 여러 클래스로 나눠지게 되어 프로젝트 구성이 복잡해질 수 있겠다는 생각이 들었습니다.

단순히 복잡도 외에 다른 단점을 더 알면 확실한 선택을 할 수 있을 텐데, 웹 서비스를 직접 개발해본 경험이 없어서 엔드포인트를 공유함으로 인해 발생할 수 있는 문제가 어떤 것이 있을지 잘 상상가지 않습니다.. 🤔 이런 상황에서 제가 고려해보면 좋을 사례가 있을까요?


리뷰해주셔서 감사합니다~! ❣️

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

jbilee committed:
feat: setup

jbilee committed:
docs: 1단계 요구사항 추가

jbilee committed:
feat(Dao): 이미 예약이 존재하는 시간을 삭제할 때 예외처리

jbilee committed:
chore: 스키마 파일 추가

jbilee committed:
feat: 예약일이 지나간 날짜인지 검증하는 로직 구현

jbilee committed:
feat: 중복된 예약인지 검증하는 로직 구현

jbilee committed:
test: 시간 생성 시 유효하지 않은 값이 입력될 때, 예약자 명, 이미 예약이 존재하는 경우 예약 시간 삭제 테스트 추가

jbilee committed:
test: 테스트에 사용되는 값 수정 및 불필요한 테스트 삭제

jbilee committed:
test: 중복 예약 검증에 대한 테스트 작성

jumdo12 committed:
refactor(FakeReservationTime):  Reservations를 컬렉션으로 변경

jumdo12 committed:
test(Reservation): 이미 지난 날짜의 경우 예약이 생성 불가능한 테스트 추가

jbilee committed:
feat: 중복 생성에 대한 ExceptionHandler 구현

jbilee committed:
feat: 삭제에 대한 ExceptionHandler 구현

jumdo12 committed:
feat: 유효하지 않은 요청 값에 대한 Exception Handler 구현

jumdo12 committed:
refactor: 응답 상태코드 수정

jumdo12 committed:
feat: theme 테이블 추가

jumdo12 committed:
docs: 2단계 README.md 내용 추가

jbilee committed:
feat: 페이지 응답 기능 구현

jbilee committed:
feat: 테마 API 전체 구현

jumdo12 committed:
refactor: Bean 애노테이션 추가

jumdo12 committed:
refactor(ReservationDao): rowMapper theme 추가

jbilee committed:
test: 테스트 코드에 테마 관련 데이터 추가

jbilee committed:
test: FakeThemeRepository 테스트 추가

jumdo12 committed:
docs: 3단계 요구사항 추가

jumdo12 committed:
feat(UserPageController): 사용자 페이지 반환 기능 추가

jbilee committed:
feat: 인기 테마 조회 API 구현

jumdo12 committed:
feat(ReservationTimeService): 예약 가능한 시간을 확인 기능 추가

jbilee committed:
chore: 완료한 TODO 삭제 및 더미 데이터 추가

jbilee committed:
refactor: 인터페이스에 구현 컨벤션 적용

jbilee committed:
test: Fake에 대한 테스트 제외

jumdo12 committed:
chore: 데이터 정렬

jumdo12 committed:
feat: theme에 따라 예약 가능한 시간조회

jumdo12 committed:
test(ReservationTimeService): 이용 가능한 시간 조회 테스트 추가

jbilee committed:
test: Theme 도메인 및 서비스 테스트

jbilee committed:
style: 코딩 컨벤션 일체화

jbilee committed:
refactor: 반환 오류 타입을 의도에 맞게 개선

jbilee committed:
refactor: ReservationDAO 쿼리문 개선

jbilee committed:
test: Reservation 서비스 관련 테스트 수정

solarsalt commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -26,21 +27,13 @@ public ResponseEntity<List<ReservationResponseDto>> getAllReservations() {
-            return ResponseEntity.badRequest().build();
-        }
+        ReservationResponseDto responseDto = reservationService.createReservation(requestDto);
+        return ResponseEntity.created(URI.create("reservations/" + responseDto.id())).body(responseDto);
```
201 status code 응답에서는 왜 URI도 함께 담아주나요? 이렇게 서버에서 URI를 응답에 담아주면, 이후에 어떤 용도로 사용되나요?

jbilee replied:
이 부분은 1차로 작업하겠습니다!

jbilee replied:
201 CREATED 상태코드는 새로운 리소스 생성 요청에 대한 응답 상태코드인데, 해당 리소스가 어느 위치에 생성되었는지는 요청자의 입장에서는 알 수 없어서 응답에 URI를 포함시킵니다. URI를 포함해서 응답하면 요청자는 응답 메세지의 location 헤더를 읽어서 전체 URI를 바로 추출해 활용할 수 있습니다.

`ResponseEntity.created()` 메서드를 사용하면 URI를 반드시 인자로 전달해야 하기도 해서, 저는 201 상태코드를 반환할 땐 URI를 포함시키는 것이 당연하다고 여기고 있었습니다. 그런데 이번에 페어 토론을 진행하면서 `현재 구현에서 클라이언트는 우리가 응답해주는 URI를 사용하고 있지 않은데, 클라이언트가 필요하지 않은 정보를 굳이 줘야 할까?`라는 크루의 의견이 있었습니다. 😮 이에 대해 저는 아래와 같이 생각했는데요,
- HTTP 표준에는 location 헤더에 URI를 포함하도록 명시되어 있는데, 애초에 이런 표준이 존재한다면 요청자가 실제로 location 값을 읽는지에 대한 여부와는 상관없이 최소한 표준에서 말하는 결과를 기대할 수 있어야 함. 표준이란 대중적인 인지도를 가진 기준이며, 여력이 된다면 표준에 맞추려고 노력하는 것이 현명하다고 판단됨
- 지금은 응답 받은 URI가 사용되지 않지만, 미래에 클라이언트의 요구사항이 바뀌거나(예: 예약완료 확인 페이지로 리다이렉트해야 할 경우) 클라이언트(우리 API의 사용자)가 다양해져서 각자의 요구사항이 달라 URI를 필요로 할 가능성이 존재. 나중에 필요해져서 추가 작업을 하는 것보다는 되도록이면 만들 때 함께 작업해두면 개발하는 입장에서 리소스적으로 유리

현재 제 의견은 표준을 따른다고 나쁠 것 없다!라고 볼 수도 있는데, 이런 의견은 설득력이 부족해 보여서 고민입니다.. 😓 해서 로이 필딩의 논문에서 말하는 과거 웹 아키텍쳐의 한계를 자세히 읽어보려고 하는데요, 이렇게 학습 시작점을 역사적인 배경부터 시작하는 것이 제 기준을 수립하는 데에 효율적일까요?

solarsalt replied:
@jbilee 
줄리가 제안해준 근거도 충분히 합리적인데요? 저는 꽤 설득되었어요~
- 미래에 리다이렉트 요구사항 추가 고려
- HTTP 프로토콜 표준을 지킬수 있으면 지키는 방향으로 일관되게 개발

로이 필딩의 논문은 REST 스타일을 제안하는 내용인데, HTTP status code 규칙 자체는 논문을 쓸 당시에도 이미 존재했어요!
그래서 HTTP 프로토콜을 최대한 따라서 URI도 헤더에 포함시킬 지는 REST와 분리해서 고민하셔도 됩니다. (물론
**REST는 메시지가 자기 서술적(self-descriptive)이어야 한다** 라는 부분이 있어서, status code를 잘 지키고 header에 정보를 포함하는 것이 자기 서술적인 메시지를 만드는 구체적인 노력이니까 REST와 완전히 무관한 것은 아니지만요😂)

> 로이 필딩의 논문에서 말하는 과거 웹 아키텍쳐의 한계를 자세히 읽어보려고 하는데요, 이렇게 학습 시작점을 역사적인 배경부터 시작하는 것이 제 기준을 수립하는 데에 효율적일까요?

- 원문에선 무엇을 주장했고, 현재의 나는 왜 이걸 지키면서 개발하는게 어려운지 알아보는 측면으로는 효율적일 수 있어요!
- 저는 notebookLM 도움으로 이런걸 질문해봤는데, 꽤 명확한 답변을 얻어서 좋았습니다. 줄리도 **궁금한 점들만 골라서 확인해본다면 전체를 다 읽는 것보다는 효율적일 것 같아요!** :) 
   - REST 아키텍처 스타일에서 URI 구문에 복수형 명사를 사용하라는 부분이 있어?
   - REST 아키텍처 스타일이 제안되기 이전의 웹 아키텍처는 무슨 문제가 있어서, 이런 제안을 한거야?
   - 여기서 말하는 resource은 정확히 어떤 의미야? 
   - `favorite/books/123`이라는 URL을 설계했어, REST 스타일에 적합한지 논문에 근거해서 평가해봐
 등등

제가 궁금한 것들을 논문에서 살펴보고나서 알게된 건 😆
- REST 논문에서는 구체적인 가이드를 준 적은 없음 (ex. URL을 정확히 어떻게 디자인 하라거나...)
- `abstraction of information, conceptual mapping, semantics`
 같은 단어가 나오는 걸 보고 왜 이렇게 논란이 많은지 이해되었음. 애초에 추상적인 규칙을 제안한거라 이후 현실에서 그 추상적 규칙을 따르는 사람들끼리 니가 맞네, 내가 맞네 하거나ㅎㅎ, 원조의 제안을 이해하지 못한 채 규칙에만 집착하는 일이 생겼구나... 
- 어지간해서는 로이 필딩 아저씨 마음에 들게 완벽하게 REST 스타일을 따르기는 어렵겠다...

solarsalt commented on src/main/java/roomescape/controller/ThemeController.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<ThemeResponseDto> postTheme(@RequestBody final ThemeCreateRequestDto requestDto) {
```
메서드명은 HTTP METHOD 이름을 따라간 것 일까요?
혹시 영어로 post ( 게시하다) 의 의미인 걸까요? 작명배경이 궁금해요~
개인적으로는 blog나 문서를 만드는 기능이라면 post가 어울리는 동사 같은데, theme를 새로하나 만들어 주는 것이라 post가 어색한 것 같아요! 줄리 의견은 어떠신가요? 여기말고 다른 컨트롤러에서도 post mapping 인 경우, post로 시작하는 이름을 붙여준 것 같더라고요~

아래는 제가 코드를 보고 들었던 질문인데요, 네이밍 의견을 정리하며 2가지도 함께 고려해보면 좋을 거 같아요!
- HTTP Method가 서버 코드의 메소드 이름에 영향을 주는 건 괜찮을까? 안좋다면 왜 안좋지?
- 어노테이션을 통해 어떤 HTTP Method 인지 알 수있는데, 이름에서 중복된 정보를 주는 것이 유용할까?

jbilee replied:
이 부분은 1차로 작업하겠습니다!

jbilee replied:
컨트롤러의 메서드명은 말씀하신대로 단순히 HTTP 메서드의 이름을 따랐는데, 제시해주신 질문을 보고 생각을 정리해봤습니다!
- HTTP 메서드가 항상 하나의 고정된 작업을 뜻하지 않음. 같은 POST 메서드여도 기존에 없던 리소스를 새로 생성하는 요청일 수도 있고, 아니면 GET의 대용으로 요청 본문에 들어있는 값을 소비하는 요청일 수도 있는데, HTTP 메서드 종류로 서버 코드의 메서드명을 그대로 짓는다면 세부적으로 어떤 작업을 수행하는 메서드인지 파악하기 어려움 🤔

말씀하신대로 post라는 표현은 테마, 시간, 또는 예약을 추가한다는 뜻으로 사용하기 어색한 것 같습니다! 해서 post 대신 add라고 표현하는 것이 더 적절할 것 같습니다.

[refactor: HTTP POST 요청을 수행하는 컨트롤러의 메서드명 개선](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/297bdcf83ea565088005e03c011c266f0297c71d)

solarsalt commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -2,27 +2,27 @@
 import java.time.LocalDate;
 
-public record Reservation(Long id, String name, LocalDate date, ReservationTime time) {
+public record Reservation(Long id, String name, LocalDate date, ReservationTime time, Theme theme) {
```
예약을 Record class로 구현하고 있네요~
record class의 원래 용도를 고려하면, 저라면 계속 record 클래스로 유지해야 할지 고민할것 같은데요.
https://docs.oracle.com/en/java/javase/21/language/records.html
https://www.baeldung.com/java-record-keyword

줄리 생각은 어떠세요? 페어의 코드라서 이렇게 시작했지만, 리팩토링 대상에 포함시킬만한 아이템 인가요?
하신다면 어떤 점을 고려해서 변경을 결정하신 것인지도 궁금하고, 안하신다면 그 이유도 궁금해요 :) 
(질문...폭탄이네요🤣)

jbilee replied:
Record 클래스는 값을 저장하는 것이 주 용도인 클래스인데, 본 애플리케이션의 기능 요구사항에 따라 구현할 비즈니스 로직은 record 클래스의 역할 범위를 넘어가는 수준의 로직이라고 생각됩니다. 다른 코멘트에서 피드백해주신 백수 객체😅에게 일을 주기 위해 서비스가 처리하던 일부 비즈니스 로직을 모델 패키지에 있는 클래스들로 옮기려고 하는데, 그렇게 되면 record보다는 일반 클래스로 보는 것이 더 적절하다고 판단됩니다.

단순히 객체의 필드값으로 null이 들어오는지를 검증하는 것을 넘어서 과거의 날짜/시간인지, 어떤 특정 문자열이 포함되어 있는지 등의 비즈니스적인 검증을 해야 하고, 추가로 record 클래스를 일반 클래스로 변경하는 작업은 페어와 작업한 코드를 몰라볼 정도로 바꾸는 수준은 아니라고 생각돼서 리팩터링 대상으로 잡으려고 합니다!

jbilee replied:
[refactor: 모델 객체의 클래스를 record에서 일반 클래스로 변경](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/4f8a336059048d265437154d11836ae37da1a585)

solarsalt commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -2,27 +2,27 @@
     }
 
-    private static void validateName(String name) {
+    private void validateName(String name) {
```
Nave, Date, reservationTime
모두 spring이 제공하는 Validaton 기능을 사용해 볼 수도 있을 부분 같은데, 적용하실 예정인가요?
줄리의 작업 계획과 결정 배경이 궁금해요~ :)

jbilee replied:
지난 미션에서 `@Validated`와 같은 애노테이션을 다른 크루들을 통해 들어는 봤지만 아직 학습해볼 기회가 없었는데, 마침 이번 미션에서 얻어갈 내용 중 일부가 검증과 관련된 부분이라고 생각해서 적용하려고 합니다!

jbilee replied:
스프링 validation 관련 기능을 DTO의 필드에 `@NotBlank`, `@NotNull` 등의 애노테이션을 추가해 사용해봤습니다!

도메인 모델이 아닌 DTO에서 체크한 이유는 컨트롤러 매개변수로 클라이언트 요청값이 들어오는 시점에 검증을 해서 내부 비즈니스 로직을 수행하기 전에 바로 클라이언트한테 시정할 부분을 알리고 작업을 (early return 하는 것처럼) 일찍이 terminate할 수 있다는 장점을 활용하기 위해서인데요, 이 과정에서 도메인 모델에 있는 validate 로직들을 삭제해야 하나 고민이 많이 들었습니다..

현재 내린 결론은 내부 비즈니스 코드의 로직적인 오류를 막기 위해서는 모델에서의 검증도 필요하다고 생각해서 삭제하지 않았는데요, validation 관련해서 주신 다른 피드백에 대한 답변을 하면서 더 생각해보려고 합니다!

[refactor: 요청 DTO에 Spring Validation 애노테이션 적용](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/3f02cc8a6e5f727e22f640b0891fa903871a6182)

solarsalt replied:
두 번 검증하는 이유도 남겨주신것 감사해요! 검증 로직을 유지하는 로직은 충분히 좋네요!

solarsalt commented on src/main/java/roomescape/domain/Theme.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    private void validateName(String name) {
+        if (name == null || name.isBlank() || name.isEmpty()) {
```
isBlank() 에서 empty도 체크해주는 것 같은데, 한번 확인 해보시면 좋을 것 같아요!

jbilee replied:
이 부분은 1차로 작업하겠습니다!

jbilee replied:
isBlank가 빈(`""`) 문자열인지를 포함해서 isEmpty에서 잡아주지 못하는 whitespace로만 이루어진 문자열까지 체크해주는 점 확인했습니다! 🤓

[refactor: 문자열 검증에서 불필요한 isEmpty 호출 제외](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/d07522ebb85234d80d927788ce43853d50fe67e1)

solarsalt commented on src/main/java/roomescape/domain/Theme.java:
```diff
@@ -0,0 +1,28 @@
+public record Theme(Long id, String name, String description, String thumbnail) {
+
+    public Theme {
+        validateName(name);
+        validateDescription(description);
+        validateThumbnail(thumbnail);
```
검증 로직을 생성시점에 하도록 결정하셨네요! 
다른 후보로는 어떤 것이 있었고, 왜 이 클래스에서 하는 것으로 결정했는지 알려주실 수 있을까요?

저는 만약에 검증로직이 많아진다면, 계속 도메인에 추가할 수 있을까? 를 고려해서
따로 validator 클래스를 만든 적이 더 많았던 것 같아요~

jbilee replied:
객체의 필드값에 대한 검증을 생성자에서 하면 검증 로직이 어디에 있는지 찾기 쉽고, 또 객체 본인이 관리하는 상태값의 유효성을 내부에서 검증하는 것이어서 객체의 책임에도 부합한다고 판단돼 생성자 검증을 선택했습니다! 만약 외부에서 검증을 한다면 RequestDTO에서 미리 검증을 해볼 수 있을텐데, RequestDTO는 뷰에 더 가까운 영역이어서 비즈니스 로직에 대한 검사라고 보기 어려울 것 같습니다.

말씀하신 Validator는 스프링 프레임워크 컨텍스트에서의 [Validator](https://docs.spring.io/spring-framework/docs/4.2.4.RELEASE/spring-framework-reference/html/validation.html)인 것 같은데 맞을까요? 👀

solarsalt commented on src/main/java/roomescape/repository/JdbcThemeDao.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    @Override
+    public List<Theme> findPopular(LocalDate start, LocalDate end) {
```
findPopular 도 좋은 이름이지만, 클래스에 있는 다른 메소드들이 파라미터를 이름에도 표현해 주는 것처럼 (ex. findById)
좀 더 좋은 이름으로 개선해볼 수도 있을 것 같아요!
쿼리를 일일이 확인하기 전 저는, 파라미터의 변수명과 쿼리결과를 보고는 'start랑 end 기간 내의 인기 테마를 조회하나보다'로 추측했거든요~ 메소드의 역할에 따라 좀 더 좋은 이름 한번 찾아볼까요?

jbilee replied:
🤔 솔라가 추측하신 내용을 듣고 고민해보니, “인기 테마”는 뷰와 비즈니스 레이어에서 쓰이는 표현이어서, 데이터베이스와 상호작용하는 DAO의 입장에서는 인기보다는 쿼리 조건문대로 “start와 end 날짜 사이에 가장 많이 예약된 테마”로 표현해야 할 것 같다는 생각도 듭니다!

영속성 계층의 관점에서는 findMostReservedByDateRange가 더 어울리는 메서드명일 것 같아서, 이렇게 수정해보겠습니다!

jbilee replied:
[refactor: JdbcThemeDao 클래스 내 인기 테마를 조회하는 메서드명 개선](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/f15b2226eff60c63afda69552e23ec66ced1b8d1)

solarsalt commented on src/main/java/roomescape/repository/JdbcThemeDao.java:
```diff
@@ -0,0 +1,99 @@
+        try {
+            return Optional.of(jdbcTemplate.queryForObject(sql, rowMapper, id));
+        } catch (EmptyResultDataAccessException e) {
+            return Optional.empty();
```
👍 항상 없을 수도 있으니 Optional 타입을 리턴하는 것은 좋아보여요~!

solarsalt commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -1,50 +1,80 @@
+    private final ThemeRepository themeRepository;
 
-    public ReservationService(ReservationRepository reservationRepository, ReservationTimeRepository reservationTimeRepository) {
+    public ReservationService(ReservationRepository reservationRepository, ReservationTimeRepository reservationTimeRepository, ThemeRepository themeRepository) {
```
줄리가 부여한 ReservationService 클래스의 책임이나 역할은 어떤 것인가요?
메소드로만 추측해보면 생성을 하고, 응답을 만들고, 삭제도 하는 것 같아요~
3가지 일을 모두 담당하게 하는 것이 적절한지 고민해보고, 리팩토링을 고려해 보시는 건 어때요?

jbilee replied:
ReservationService는 사용자 예약(Reservation 모델)에 대한 데이터의 CRUD 작업을 한곳에서 처리하는 모듈로 구현했습니다. Theme 요구사항이 생기기 전까지는 느끼지 못했는데, ThemeRepository도 의존성으로 추가하게 되면서 한 클래스가 알아야 하는 repository가 많은 것 같다는 생각이 들었습니다.. 게다가 정작 ReservationTimeRepository와 ThemeRepository는 예약 객체 생성에 필요한 값들을 관계형 테이블에서 읽어오는 데에만 사용돼서, ReservationService의 역할에서 분리해볼만 한 것 같습니다. 🤔

Create, read, delete와 같이 컨트롤러가 서비스에게 처리해달라고 요청하는 메서드 호출은 그대로 두되, 해당 행동을 수행하는 과정에서 일부 작업을 다른 객체에게 위임하는 방향으로 리팩터링하면 어떨까 싶습니다. 👀

jbilee replied:
ReservationService 클래스가 컨트롤러한테 전달 받는 RequestDto를 통해 DTO를 이미 알고 있어서, ResponseDto 응답을 만드는 책임도 함께 갖는게 자연스럽다고 생각돼서 응답 생성에 대한 역할을 따로 분리하는 구조가 잘 그려지지 않는 것 같습니다.. 😵
해서 우선 ReservationService에서 수행하던 Reservation 객체 생성을 별도의 클래스에서 처리하도록 수정해봤습니다!

[refactor: ReservationService 역할 분리](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/46dc4e5a7258fd2ba81b435e2deb199c774eff8f)

solarsalt replied:
관련한 코멘트는 새로 작업해주신 팩토리 클래스에 남겼습니다 :)

solarsalt commented on src/main/java/roomescape/repository/JdbcThemeDao.java:
```diff
@@ -0,0 +1,99 @@
+        try {
+            return jdbcTemplate.update(sql, id);
+        } catch (DataIntegrityViolationException e) {
+            throw new IllegalStateException("[ERROR] 이 테마는 이미 존재합니다. id : " + id);
```
예외를 전환하셨네요!
만약에 DataIntegrityViolationException 발생했을 때, 개발자가 stackTrace를 보거나 이 예외가 발생했다는 것을 감지할 수있는지 점검해보면 좋겠어요 :)

jbilee replied:
현재 구현돼있는 코드로 예외를 발생시켜봤더니 터미널의 오류 stack trace에 IllegalStateException 클래스와 제가 지정한 오류 메세지 외 다른 힌트가 없어서, 정확히 어떤 이유에서 예외가 발생했는지 알 방법이 없었습니다. 😅

관련해서 찾아보니 예외 객체를 생성할 때 두번째 매개변수로 Throwable 객체를 전달하면 stack trace에 예외 발생의 근원지로 로깅된다는 점을 알게 됐습니다! 본래 발생하는 예외를 catch해서 다른 예외로 다시 throw할 땐 cause를 인자에 포함시켜서 stack trace에서 파악이 가능하도록 개선하려 합니다.

다만 마지막으로 throw된 예외 클래스와 cause 예외 클래스까지만 파악하기 쉽고, 중간 단계에 throw되는 다른 예외들은 stack을 자세히 확인해야 알 수 있어서, 예외 전환을 해야 하는 경우에는 더 의미있는 예외 클래스로 throw해야겠다는 생각이 들었습니다. 🤓

jbilee replied:
IllegalStateException 예외 클래스에 대한 피드백을 반영하면서 DAO가 예외를 전환해 다시 throw하는 본 로직이 최종적으로는 없어졌지만 기록 차원에서 커밋 남겨드립니다!

[refactor: 예외 전환 시 기존 예외를 cause로 전달하도록 수정](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/34f7687c77b92b61a5e9de714d31b0c7b0a7fcfc)

solarsalt commented on src/main/java/roomescape/controller/ReservationTimeController.java:
```diff
@@ -25,23 +33,21 @@ public ResponseEntity<List<ReservationTimeResponseDto>> getAllReservationTimes()
-            return ResponseEntity.notFound().build();
-        }
+        reservationTimeService.deleteReservationTimeById(id);
+        return ResponseEntity.noContent().build();
```
noContent 는 어떤 status code를 응답하고, 왜 여기에서 사용하기 적절한가요?

jbilee replied:
No content는 204 상태코드를 응답하며, “요청자의 작업이 정상적으로 처리됐고 응답 본문에 넣어 보내줄 컨텐츠가 없음”을 뜻합니다. No content는 HTTP 표준에서 PUT이나 DELETE 요청에 대한 작업이 성공적으로 완료되었을 때 보내는 상태코드인 것으로 구분되어 있지만, 설계에 따라 204가 아닌 200 OK를 반환하기도 합니다.

200와 204를 굳이 구분하는 이유는 처리 상태를 호출자에게 알려줘야 할 때와 그럴 필요가 없는 상황을 구분하기 위함이라고 생각합니다. 예를 들어 사용자 계정 삭제처럼 테이블에서 완전히 지우는 대신 데이터가 더이상 유효하지 않다는 의미의 값을 별도의 column에 저장하도록 설계한 경우, 삭제 작업의 처리 여부를 200 OK 응답의 본문에 담아 호출자에게 알릴 수 있습니다. 이외에도 삭제한 데이터의 일부 원본 값을 응답 본문에 포함시켜 호출자에게 알려줘야 할 때 204 대신 200을 사용하는 것이 적합할 것 같습니다!

ReservationTime은 사용자 계정처럼 처리 상태를 따로 알아야할 만큼 중요하지 않고, 아무때나 추가 혹은 삭제해도 문제없는 정보여서 클라이언트에게 따로 처리 상태를 알려줄 필요가 없다고 여겨 204 상태코드를 사용하였습니다. 👀

solarsalt commented on src/test/java/roomescape/repository/FakeThemeRepository.java:
```diff
@@ -0,0 +1,102 @@
+                .map(Map.Entry::getKey)
+                .toList();
+
+//        List<Reservation> reservations1 = reservations.stream()
+//                .filter(reservation -> reservation.date().isBefore(end) && reservation.date().isAfter(start))
+//                .toList();
+//
+//        for (Reservation reservation : reservations1) {
+//            counts.computeIfPresent(reservation.theme().id(), (k, v) -> v + 1);
+//            counts.putIfAbsent(reservation.theme().id(), 1);
+//        }
+//
+////        counts.entrySet().stream().sorted(Comparator.comparing((entry1) -> ));
```
주석 코드는 과감하게 삭제하는 것이 더 읽고 관리하기 좋아요~!

jbilee replied:
다급한 마음에 미션 제출 전 마지막 점검을 프로덕션 코드 위주로만 하느라 놓친 것 같습니다... 😭 캐치해주셔서 감사합니다, 바로 삭제하겠습니다!!

jbilee replied:
[refactor: 불필요한 주석 코드 삭제](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/04e47e4bf94bbbe4abafc874ab818d72dd213ca9)

solarsalt commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -1,50 +1,80 @@
-        Reservation requestReservation = dto.createWithoutId(reservationTime);
+                .orElseThrow(() -> new NotFoundException("[ERROR] 예약 시간을 찾을 수 없습니다. id : " + dto.timeId()));
 
+        validateDuplicate(dto.date(), reservationTime.startAt());
+        validateReservationDateTime(dto.date(), reservationTime.startAt());
```
똑같은 파라미터를 받아서 로직을 수행하는 메소드가 2개네요.
리팩토링 시그널일 수도 있겠는데요?

jbilee replied:
이 부분은 1차로 작업하겠습니다!

jbilee replied:
createReservation 메서드를 전체적으로 리팩터링하면서 두 검증 메서드가 하나로 통합되었습니다! 같은 파라미터를 받는 경우를 리팩터링 시그널로 생각해보지 못했는데, 앞으로 rule of thumb으로 활용해보겠습니다! 🤓
(코멘트 두개에 대한 작업을 한 커밋에 했는데, 커밋 링크는 혼선이 없도록 아래 createReservation 메서드 리팩터링 관련 코멘트에만 달아두겠습니다!)

solarsalt commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -1,50 +1,80 @@
+        }
+    }
+
+    private void validateReservationDateTime(LocalDate date, LocalTime time) {
+        LocalDateTime dateTime = LocalDateTime.of(date, time);
+        LocalDateTime currentTime = LocalDateTime.now();
+        if (dateTime.isBefore(currentTime) || dateTime.equals(currentTime)) {
+            throw new InvalidRequestException("[ERROR] 현 시점 이후의 날짜와 시간을 선택해주세요.");
+        }
```
` validateReservationDateTime(dto.date(), reservationTime.startAt());`
dto객체에서 특정한 데이터만 꺼내서(사실상 getter를 쓰신거죠~ ㅎㅎ) 검증하는 로직이 service class의 private method로 존재하네요! 저 역시 바쁘게 구현하다보면, 정신 차려보면 이런 메소드가 생겨있어요~

getter를 쓰는 자체가 문제는 아니고요, getter를 자꾸 쓰면 객체가 아무할 일 없는 백수가 되는 경향이 있는 것을 주의하기만 하면 됩니다! 그리고 이렇게 발견하면 리팩토링을 고민하면 되고요!
좀 더 객체가 직접 일하게 하는 방식으로 리팩토링 해보면 어떨까요?

jbilee replied:
이 부분은 1차로 작업하겠습니다!

jbilee replied:
Reservation 객체가 날짜와 시간에 대한 정보를 갖고있어서 예약일이 과거의 날짜인지에 대한 확인을 Reservation에서 수행하고, 그 결과에 따라 예외를 반환하는 것을 서비스에서 처리하도록 수정해봤습니다!

[refactor: 지난 예약 날짜인지 확인하는 로직을 Reservation 클래스에서 수행하도록 수정](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/bf9f8679a85523479e9a84ff5f169486ed34211e)

solarsalt commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -1,50 +1,80 @@
                 .orElseThrow(() -> new IllegalStateException("[ERROR] 알 수 없는 오류로 인해 예약 생성을 실패하였습니다."));
 
-        return ReservationResponseDto.from(newReservation, newReservation.time());
+        return ReservationResponseDto.from(newReservation, newReservation.time(), theme);
```
이 메소드가 전체적으로 line 수가 많은 편이에요! 리팩토링 시그널입니다~
차근차근 개선해보면 어떨까요?! 한번에 우아하게 가지 않아도 괜찮으니, 현재 상태보다 좀 더 간결한 메소드로 바꿔보시죠!

jbilee replied:
이 부분은 1차로 작업하겠습니다!

jbilee replied:
기존에는 ReservationTime과 Theme을 순차대로 쿼리하고 도중에 읽어오는 값으로 유효성을 먼저 검증하느라 메서드를 분리하기 어렵고 가독성이 떨어졌는데, 리팩터링을 하면서 아래와 같은 플로우로 로직의 순서를 바꿔보았습니다
  - 변경 전: ReservationTime 조회 → 중복 예약인지 검증 → 지난 날짜인지 검증 → Theme 조회 → Reservation 객체 생성 → DB에 저장
  - 변경 후: ReservationTime 조회 → Theme 조회 → Reservation 객체 생성 → 중복 예약 및 지난 날짜인지 검증 → DB에 저장

[refactor: ReservationService 클래스 내 createReservation 메서드 리팩터링](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/c1c2344bc44681359a4d37a30ef4fb94b4475bfe)

solarsalt commented on src/main/java/roomescape/service/ReservationTimeService.java:
```diff
@@ -1,42 +1,70 @@
-        return ReservationTimeResponseDto.from(savedTime);
+            return ReservationTimeResponseDto.from(savedTime);
+        } catch (IllegalStateException e) {
+            throw new DuplicateContentException(e.getMessage());
```
다른 곳에서는 최종 Exception을 IllegalStateException로 하는 경우가 많았는데,
여기서는 IllegalStateException를 DuplicateContentException이라는 custom 예외로 전환해주셨네요!

처음 읽는 저는 이 서비스가 `IllegalStateException`을 어떤 때는 커스텀 예외로 바꿔주고, 어떤 때는 최종 예외로 사용하는 것인지 기준을 알기가 어려웠어요~ 

전체적으로 커스텀 예외를 사용하는 방향으로 설계하셨다면, IllegalStateException 을 최종 예외로 던졌던 경우를 해결해보면 좋을 것 같은데요. 제가 생각해본 방식은 크게 2가지 인데, 좀 더 고민하면 다른 선택도 물론 가능하고요~ 
1. IllegalStateException로 처리했던 경우마다 적절한 커스텀 예외를 만들어 주는 방식
2. IllegalStateException로 처리한 경우를 새로운 이름의 커스텀 예외 1개로 대체하는 방식

저의 제안은 참고로만 봐주시고, 선택은 줄리에게 맡겨둘게요!
이 부분은 다른 크루들과 💬얘기해본 뒤에 예외를 개선 보시는 것도 도움이 될거같아요!

jbilee replied:
그동안 피드백 주셨던 부분들과 크루들과 대화해본 내용들을 참고해 IllegalStateException에 대한 생각을 정리해봤습니다!

- DAO에서 발생하는 예외들을 IllegalStateException 하나로 표현하기에는 상황이 근본적으로 다르거나 서비스와 중복되는 경우도 존재 (중복으로 인해 심지어 서비스에서 또다시 다른 예외로 이중 래핑하는 경우도 발생..)
- IllegalStateException으로 전환한 이유는 서비스가 DAO에서 발생한 상세 문제를 서비스의 관심사에서 제외시키기 위함이었는데, DAO에서 발생한 문제를 서비스가 알아야 엔드 유저에게 어떤 메세지를 보여줄지를 알맞게 처리할 수 있어서 모순이 있음
- 파악하기 어려운 예외 클래스로 래핑하지 않고, DAO의 예외를 서비스가 catch해서 비즈니스 맥락에서의 예외로 변환하는게 적절할 것 같음
- IllegalStateException은 CRUD 작업에 대한 상태를 포괄적으로 표현하기 위함이었는데, 이건 상태보다는 처리결과에 가까우며 실제로 IllegalStateException이 말하는 [상태]와는 다르다고 판단됨
- IllegalStateException은 인프라의 시스템적인 오류로 create 작업이 실패할 때, "현재 해당 작업을 수행할 수 있는 상태가 아니다"를 나타내는 예외로 사용하기로 결정

위와 같은 이유로 DAO에서 IllegalStateException을 throw하던 코드를 삭제하고 서비스가 DAO의 예외를 직접 판단하도록 수정했습니다!

[refactor: IllegalStateException 사용 기준 변경](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/810874cfa77f8c42391a0aff50ded5105ffc07ae)

solarsalt commented on src/main/java/roomescape/service/ReservationTimeService.java:
```diff
@@ -1,42 +1,70 @@
-                .orElseThrow(() -> new IllegalStateException("[ERROR] 알 수 없는 오류로 인해 예약시간을 생성 실패하였습니다."));
+        try {
+            ReservationTime savedTime = reservationTimeRepository.save(requestTime)
+                    .orElseThrow(() -> new IllegalStateException("[ERROR] 알 수 없는 오류로 인해 예약시간을 생성 실패하였습니다."));
```
예약 메시지도 상황에 맞게 대부분 열심히 작성해주신 것 확인했어요~!
이 경우에는 약간 아쉬운 에러메시지 같아요. 이 메시지 그대로 유저가 보는 것을 의도하셨는지도 궁금합니다~!
유저에게 어떤 에러 메시지가 더 도움이 되고, 명확할 지를 고민해서, 에러메시지도 수정해 볼 수 있을 것 같네요.

jbilee replied:
해당 오류 메세지가 어떤 상황에서 발생하는지를 다시 확인해보니 POST 요청으로 들어온 데이터를 생성한 후 generate된 id로 다시 SELECT문을 쿼리해 Optional<> 객체를 반환하는 구조임을 깨달았습니다. DB에 create 직후 다시 read했을 때 데이터가 없다면 그 원인을 파악하기 어려워 “알 수 없는 오류”로 표현하게 됐는데, 사용자 입장에서는 이해할 수 없는 표현이기에 결과와 대처 방안에 초점을 맞춘 메세지가 필요할 것 같습니다.

JdbcTemplate이 직접적인 예외를 반환하지는 않았지만 생성한 id로 데이터가 조회되지 않는다면 상황을 내부 시스템상 오류로 인한 "예약 실패"로 보고, 사용자에게는 나중에 다시 시도하거나 고객센터/관리자에게 문의하도록 유도하는 것이 어떨까 싶습니다. 🤔

> 예약시간을 저장할 수 없습니다. 잠시 후 다시 시도해 주세요.
> 예약시간을 저장할 수 없습니다. 관리자에게 문의해 주세요.

제가 사용자 입장에서 각종 서비스를 사용할 때 위와 같은 표현을 자주 접해서 고려해보게 됐습니다. 현재 구조에서는 작업이 실패하는 이유가 예기치 못한 인프라의 문제일 것으로 예상하는데, 그런 경우 단기간 안에 해결하지 못할 가능성이 커서 위 메세지 둘 중 하나를 선택한다면 문의로 유도하는 것이 나을 것 같습니다. 시간이 해결해주지 못할 수도 있는데 사용자에게 시도를 권장하게 되어서요. 😨

jbilee replied:
[refactor: 리소스 생성 실패에 대한 오류 메세지 개선](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/265f4950abe83bc9196eb81f8b096ea7f2e5029b)

solarsalt commented on src/main/java/roomescape/domain/Theme.java:
```diff
@@ -0,0 +1,28 @@
+        }
+    }
+
+    private void validateThumbnail(String thumbnail) {
```
thumbnail이 url 형식인지도 검증해야 할까요?
어떤 기준을 고려해서 검증 여부를 결정하실 지 궁금해요~!

jbilee replied:
name은 null값 뿐만 아니라 최소한 한글자 이상인지까지도 확인하는데, thumbnail은 이미지 URL이라는 아주 특정한 형식의 문자열을 저장하는 필드면서 비즈니스적으로 아무런 검증이 없는 것이 의문스럽습니다! 😂

지금 구현에서는 외부에 업로드되어 있는 이미지 주소를 입력 받도록 되어있어서, thumbnail 값이 `http://` 또는 `https://`로 시작하는지를 검증하려고 합니다.

처음에는 흔히 허용되는 이미지 파일 확장자로 끝나는지까지 확인하려고 했는데, 최근에는 파일 확장자가 없는 주소도 이미지일 수 있어서 제외하기로 했습니다. 👀

jbilee replied:
[feat: Theme 클래스에 thumbnail 값 검증 조건 추가](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/8b0dc1971d12e6a6a1f72a07d30b306b47508be5)

solarsalt replied:
**특정한 형식의 문자열을 저장하는 필드면서 비즈니스적으로 아무런 검증이 없는 것이 의문** 이라는 결론을 내리신 것을 존중합니다!
어떻게 검증할 지도 일단 가능한 심플한 방식부터 시작한 것도 좋아요~

solarsalt commented on src/main/java/roomescape/repository/JdbcThemeDao.java:
```diff
@@ -0,0 +1,99 @@
+            long id = jdbcInsert.executeAndReturnKey(params).longValue();
+            return findById(id);
+        } catch (DuplicateKeyException e) {
+            throw new IllegalStateException("[ERROR] 이미 등록된 테마 입니다.");
```
IllegalStateException 는 어떤 상황에서 사용하는 것을 의도하셨나용? 
코드 전반에서 자주 사용되길래, 어떤 의미를 염두에 두고 적용했는지 궁금했어요~

jbilee replied:
이 부분은 1차로 작업하겠습니다!

jbilee replied:
IllegalStateException은 영속성 계층에서 수행하는 CRUD 작업의 상태를 나타내기 위해 사용했습니다. JDBC로 데이터베이스를 조작하다가 발생하는 오류는 비즈니스 로직의 관심사가 아니라고 판단돼서 DAO의 메서드에서 발생하는 EmptyResultDataAccessException, DataIntegrityViolationException 등의 예외를 IllegalStateException로 한번 래핑한 다음 서비스 계층에 throw하도록 구현했는데, 래핑이 필요한 상황인지를 판단해야 해서 기능을 추가하거나 유지보수할 때 코드를 읽는 비용이 높아지고, 해당 코드를 처음 보는 사람의 입장에서는 직관적이지 않은 것 같아 고민입니다.. 😥

IllegalStateException 관련해서 주셨던 다른 피드백을 2차 수정 때 작업하면서 어떻게 쓰는게 적합할지에 대해 더 생각해보려고 합니다!

solarsalt replied:
DAO의 메서드에서 발생하는 DB관련 에러를 IllegalStateException로 래핑하신거군요!
의도는 충분히 이해했습니다.
`IllegalStateException` 가 모호한 이름처럼, 구체적인 정보를 주지 않으면서 && 애매모호한 느낌이라 개인적으로는 의도적으로 사용해본 경험이 거의 없어요. 

다시 설명해 주신 의도로 돌아가본다면, DB관련 에러를 래핑하는 커스텀 예외를 만드는 것도 가능하겠네요~
물론 고민해주신 것처럼 모든 DB관련 예외의 래핑이 필요한지? 언제 래핑이 필요한지? 좀 더 고민해보시는 것도 좋습니다.

jbilee replied:
이번에 IllegalStateException을 여러군데 사용해보면서 말씀하신 애매모호함을 뼈저리게 느낄 수 있었던 것 같습니다. 😂

고민한 결과 DB에서 발생하는 예외를 래핑하는 것은 명확한 장점을 보이지 않는 이상 안하는 편이 낫겠다는 생각을 갖게 됐습니다. 오류를 추상화시켜도 정확한 원인 파악은 여전히 할 수 있어야 하는데, 그러면 결국 새 예외 객체를 반환할 때 cause도 함께 전달해줘야 하고, 그러면 예외의 클래스명만 새단장시키는 것 같다는 생각이 들었습니다. 대신 DataIntegrityViolationException처럼 눈에 잘 들어오지 않는 복잡한 이름이라면 더 간결한 예외로 래핑해줘도 좋을 것 같습니다!

solarsalt commented on src/main/java/roomescape/controller/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,31 @@
+import roomescape.exception.NotFoundException;
+
+@RestControllerAdvice
+public class GlobalExceptionHandler {
```
이것도 controller 패키지에 둘지, exception 패키지에 둘지 고민해볼만 한 것 같아요!
저는 예외를 핸들링하는 것이 이 클래스에 역할 같아서, exception에 두는 것이 덜 헷갈릴 것 같은데, 줄리는 어때요?

jbilee replied:
저도 동의합니다! GlobalExceptionHandler가 controller 패키지에 위치해야 할 만큼 컨트롤러와 연관성을 갖고 있지 않아서, exception 패키지에 속해있는 것이 해당 클래스의 path를 보았을 때에도 더 직관적일 것 같습니다. 😀

jbilee replied:
[chore: GlobalExceptionHandler 클래스 패키지 이동](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/54f9ed605f0ab02561a54c0a9e5061fcca46ec44)

solarsalt commented on src/main/java/roomescape/controller/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,31 @@
+        return new ResponseEntity<>(ex.getMessage(), HttpStatus.NOT_FOUND);
+    }
+
+    @ExceptionHandler(IllegalStateException.class)
+    public ResponseEntity<String> handleIllegalStateException(IllegalStateException ex) {
+        return new ResponseEntity<>(ex.getMessage(), HttpStatus.BAD_REQUEST);
+    }
```
IllegalStateException를 던진 모든 코드에서 `BAD_REQUEST`상태코드로 응답하기 적절한지 검토해보면 어떨까요?
만약 전체적으로 IllegalStateException의 사용을 개선하신다면, 더 적합한 상태코드 후보를 찾아보시는 것도 추천합니다.

예를 들면 아래의 경우도 BAD_REQUEST가 적절한가용?
```
Theme savedTheme = themeRepository.save(requestTheme)
                    .orElseThrow(() -> new IllegalStateException("[ERROR] 알 수 없는 오류로 인해 테마를 생성 실패하였습니다."));
```

jbilee replied:
확실히 하나의 IllegalStateException으로 상태코드를 일괄처리하기에는 무리가 있어보입니다! IllegalStateException을 사용하던 구조를 바꾸면서 적절한 상태코드를 응답하도록 수정하려 합니다!

jbilee replied:
HTTP 상태코드에 대한 각종 소스를 참고하면서 총 4가지 상태코드를 사용하는 방향으로 수정해봤습니다!

- 400 BAD REQUEST: 요청 본문의 값이 잘못된 포맷이거나 비즈니스 요구사항을 충족하지 못했을 때, 생성 요청 데이터가 중복 값일 때
- 409 CONFLICT: 다른 데이터베이스 테이블과 외래키로 연결돼있는 데이터를 삭제하려고 할 때
- 404 NOT FOUND: DB에 없는 id의 데이터를 조회하려고 할 때
- 503 SERVICE UNAVAILABLE: 애플리케이션이 예기치 못한 이유로 정상 동작하지 않을 때 (IllegalStateException을 이 상태코드로 처리합니다)

[refactor: IllegalStateException 사용 기준 변경](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/810874cfa77f8c42391a0aff50ded5105ffc07ae)

solarsalt reviewed (changes requested):
줄리~ 리뷰 분량 조절에 실패했어요😅 쏘리~
연휴 동안 찬찬히 개선해보시면 어떨까요?
한번에 개선이 어렵다면, 1차로 개선해서 제출해주시고
그 다음턴에 작업하시는 방식도 괜찮으니 편하게 작업 계획을 알려주세요~

> Q. 구현한 컨트롤러 중 테마 API와 예약 시간 API의 컨트롤러에는 한 엔드포인트에 각각 다른 HTTP 메서드가 매핑되어 있습니다. 예를 들어서 themes/popular에 GET 요청을 보내면 인기 테마 목록을 반환하는 서비스 메서드를 호출하고, themes/popular에 DELETE 요청을 보내면 popular라는 id를 가진 리소스를 삭제하는 서비스 메서드를 호출하는 구조입니다.
> 
> 올바른 메서드로 요청을 보내면 문제될 것이 없지만, 한 엔드포인트가 두가지 일을 하는 것으로 보여 고민하게 되었습니다. 그렇다고 컨트롤러를 분리하면 동일한 theme 모델에 대한 컨트롤러인데 여러 클래스로 나눠지게 되어 프로젝트 구성이 복잡해질 수 있겠다는 생각이 들었습니다.
> 
> 단순히 복잡도 외에 다른 단점을 더 알면 확실한 선택을 할 수 있을 텐데, 웹 서비스를 직접 개발해본 경험이 없어서 엔드포인트를 공유함으로 인해 발생할 수 있는 문제가 어떤 것이 있을지 잘 상상가지 않습니다.. 🤔 이런 상황에서 제가 고려해보면 좋을 사례가 있을까요?

제가 줄리의 질문을 잘 이해한 것이 맞는지 확인해주세요~

> 예를 들어 ThemeController처럼
> `/themes` 라는 동일한 URL 이 있는데, GET인지 POST인지만 다르면서 완전히 다른 로직을 수행하는 것이 문제가 없는지? 코드에서 혹시 클래스 분리를 해야하는지? 
가 맞을까요?

일단 REST API를 공부하셨다면, 이미 답을 찾으셨을 것 같지만
그게 바로 필딩이 REST를 만들면서 원하던 바입니다! 그래서 보통은 http method별로 class를 분리해주지 않아요~

HTTP Method가 요청의 type을 이미 표현하게 하고 있으니,
URL은 동일하게 사용하면서 
각 요청에서 일관성&확장성 있게 쓸 수 있게 하는 것이 REST 스타일에서 제안하는 바라고 생각해요.

커뮤니케이션 비용을 줄이면서도 http method 에 정보가 있으니, URL에 중복으로 담을 필요는 없다고 생각해보면 어떨까요? (ex. 테마 생성 URL을 REST를 딱히 고려하지 않고 설계한다면 온갖것이 가능해져요 /themes/post , theme/doPost, themes/create, themes/new .... 개발자마다 모두 다른 방식으로 커뮤니케이션 했을 때의 혼란과 비용을 줄이는 스타일을 제안한거죠!) 

웹 세계는 비용을 고려하면서도, 자세히보면 합리적인 방식의 프로토콜(약속)들이 많아서 아주 재밌어요!
예를 들면 패킷도 데이터를 쪼개서 정해진 길도 없이, 보내는 데도 결국 목적지에 도착하거든요~
HTTP 나 REST 아키텍처도 그런 배경에서 세부적인 규칙이 정해진 것들이 많아요! 

지금처럼 어색한 규칙은 그냥 외우지 않고, 이렇게 의문을 가지면서 계속 질문하는 것도 너무 좋아요!!👍🏻

- 코드레벨 말고 테스트 중에 아쉬운 점을 발견해서 제안해 봅니다.
혹시 커스텀 예외마다 열심히 작성해주신 에러 메시지가 사용자가 보는 것까지도 의도하셨다면, 서비스 페이지에서도 유저가 볼 수 있어야 할 것 같아요! 개발자 도구를 켜지 않는 이상 에러메시지를 확인할 수 없어서, 사용자 입장에서는 원하는 액션이 왜 실패하는 지 알수 없어서 여러번 버튼을 눌러보는 시도밖에 할 수있는 게 없었어요 😭
JS에서 서버에서 받은 에러 메시지를 alert으로 보여주는 것이 그리 어렵지 않으니, 적용한다면 훨씬 높은 퀄리티의 서비스를 제공할 수 있을 거에요!

<img width="1417" alt="스크린샷 2025-05-02 오후 9 21 04" src="https://github.com/user-attachments/assets/0f504ff5-8a44-440c-a454-82715cf14e3a" />

jbilee committed:
refactor: 불필요한 주석 코드 삭제

jbilee committed:
refactor: RowMapper에서 생성한 인스턴스를 변수에 초기화하지 않고 바로 반환하도록 개선

jbilee committed:
refactor: 모델 객체의 클래스를 record에서 일반 클래스로 변경

jbilee committed:
refactor: 문자열 검증에서 불필요한 isEmpty 호출 제외

jbilee committed:
refactor: JdbcThemeDao 클래스 내 인기 테마를 조회하는 메서드명 개선

jbilee committed:
refactor: 지난 예약 날짜인지 확인하는 로직을 Reservation 클래스에서 수행하도록 수정

jbilee committed:
refactor: ReservationService 클래스 내 createReservation 메서드 리팩터링

jbilee committed:
feat: Theme 클래스에 thumbnail 값 검증 조건 추가

jbilee committed:
fix: ReservationService 클래스 내 잘못된 메서드 호출 수정

jbilee committed:
fix: 예약 중복 확인 로직 관련 버그 수정

- 같은 날짜와 시간만으로 중복을 구분하던 부분을 같은 날짜, 시간, 테마로 중복 처리를 하도록 수정

jbilee committed:
test: 리팩터링으로 인해 발생한 로직 변동사항을 테스트 코드에 반영

jbilee committed:
refactor: HTTP POST 요청을 수행하는 컨트롤러의 메서드명 개선

jbilee committed:
refactor: 리소스 생성 실패에 대한 오류 메세지 개선

jbilee committed:
fix: FakeReservationRepository 메서드 내 잘못된 객체 비교 수정

jbilee committed:
refactor: ReservationTimeController 인스턴스 필드에 final 적용

jbilee committed:
refactor: 요청 DTO에 Spring Validation 애노테이션 적용

jbilee committed:
test: 요청 시간 포맷에 대한 테스트 삭제

- LocalDate으로 파싱만 가능한 형태라면 요청 입력의 포맷은 문제가 없다고 판단되어 관련 제약사항 삭제

jbilee committed:
refactor: ReservationService 역할 분리

jbilee committed:
test: ReservationService 클래스 리팩터링에 따른 테스트 코드 수정

jbilee commented:
@solarsalt 솔라~ 1차로 작업하기로 했던 코멘트에 모두 댓글을 달아두었습니다! 🌞 혼선방지를 위해 나머지 피드백은 1차 리뷰가 끝나실 때까지 댓글을 안 달고 별도로 트랙킹하면서 작업하고 있겠습니다!

> 제가 줄리의 질문을 잘 이해한 것이 맞는지 확인해주세요~

맞습니다!! 아직 REST API가 달성하려는 목적을 전부 숙지하지 못해서 솔라의 피드백이 정말 많은 도움이 되었어요! 🙏 HTTP 메서드가 불필요한 중복을 제거해주는 중요한 역할도 겸하고 있다는걸 예시로 주신 엔드포인트들을 보고 실감할 수 있었습니다. 😂

추가로 짚어주신 HTML 뷰의 개선사항도 적용해두겠습니다! :)

solarsalt commented on src/main/java/roomescape/dto/ThemeCreateRequestDto.java:
```diff
@@ -0,0 +1,17 @@
+        @NotBlank(message = "테마의 이름은 1글자 이상으로 이루어져야 합니다.") String name,
+        @NotBlank(message = "테마 설명이 없습니다.") String description,
+        @NotBlank(message = "테마 이미지가 없습니다.")
+        @URL(message = "URL 형식이 아닙니다.") String thumbnail
```
@URL 어노테이션을 사용해보셨네요! 좋은 시도에요 👍 
이 어노테이션에서 검증하고 싶은 속성도 직접 지정해주면 좋을 것 같아요!
(ex.protocol = "https")

jbilee replied:
protocol을 지정해주지 않으면 ftp 등의 다른 프로토콜도 허용하는군요! 😳 지정해주는게 좋을 것 같습니다!

jbilee replied:
[refactor: URL 애노테이션에 protocol 속성 명시](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/aa3c25ac3dde4c138c95ce921f28e4817ef917e5)

solarsalt commented on src/test/java/roomescape/service/ReservationServiceTest.java:
```diff
@@ -63,11 +72,65 @@ void createReservationTest() {
+            ReservationCreateRequestDto requestDto = new ReservationCreateRequestDto("가이온", LocalDate.now().plusDays(7), 1L, 1L);
 
-            assertThatThrownBy(() -> reservationService.createReservation(requestDto)).isInstanceOf(IllegalStateException.class);
+            assertThatThrownBy(() -> reservationService.createReservation(requestDto));
```
어떤 예외가 발생하기를 기대하는지 누락되었어요~

jbilee replied:
테스트에 예외 클래스 체킹 로직을 추가했습니다! 🙏

[test: ReservationService 테스트에 누락된 예외 클래스 검증 추가](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/7df3daa138159fa6d3144d547259ff14ed8e4990)

solarsalt commented on src/main/java/roomescape/service/ReservationRequestFactory.java:
```diff
@@ -0,0 +1,52 @@
+        this.themeRepository = themeRepository;
+    }
+
+    public Reservation createReservationRequest(ReservationCreateRequestDto dto, ReservationRepository reservationRepository) {
```
ReservationRepository은 파라미터로 넘겨주지 않아도 이 클래스에서 사용할 수 있어요~
이유가 뭘까요?

jbilee replied:
ReservationRequestFactory와 ReservationRepository는 둘 다 스프링 빈이어서 제가 직접적으로 파라미터로 넘겨주지 않아도 스프링이 ReservationRepository를 대신 주입시켜줍니다! 결국 ReservationRequestFactory도 어느 시점에서는 로직에 ReservationRepository 객체를 사용하는데, 단순한 맥락에서의 분리만 생각하다가 로직적 모순에 빠져버린것 같습니다 😂 ReservationRequestFactory를 전체적으로 개선하면서 불필요한 주입을 제외시켜보겠습니다!

jbilee replied:
ReservationRequestFactory를 BookService로 바꾸면서 repository를 메서드 파라미터에서 삭제하였습니다!

solarsalt commented on src/main/java/roomescape/service/ReservationRequestFactory.java:
```diff
@@ -0,0 +1,52 @@
+@Component
+public class ReservationRequestFactory {
+
+    private final ReservationTimeRepository reservationTimeRepository;
+    private final ThemeRepository themeRepository;
```
ReservationRepository까지 이 클래스에서 사용하고 있으니 
오히려 이 클래스 역시 제2의 ReservationService 가 된 것 같기도 합니다 ㅠㅠ 한번 개선해볼까요?

jbilee replied:
😭 스프링 빈 본연의 기능을 활용하지 못하고 어색하게 분리해버린 것 같아요! Factory는 BookService로 만들어서 예약 생성과 삭제를 수행하도록 수정해봤습니다!

[refactor: ReservationRequestFactory를 서비스 객체로 변경](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/809a1b2ed6c1ff98e150ee12291fb070c36e9328)

solarsalt commented on src/main/java/roomescape/service/ReservationRequestFactory.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    private void validateReservationRequest(Reservation reservationRequest, ReservationRepository reservationRepository) {
+        if (reservationRequest.isBeforeCurrentDateTime()) {
+            throw new InvalidRequestException("[ERROR] 현 시점 이후의 날짜와 시간을 선택해주세요.");
+        }
```
ReservationTime 이나 Theme를 DB에서 조회하기 전에, 이 로직을 수행하면 어떤 장점이 있을까요?

jbilee replied:
ReservationTime, Theme을 데이터베이스에서 불러오기 위해서는 SQL 쿼리문을 실행시켜야 합니다. 데이터베이스 테이블의 내용을 읽는 비용이 두번이나 발생하게 되는데, 사전에 isBeforeCurrentDateTime 메서드로 유효한 날짜의 예약을 생성할 수 있는지를 먼저 검증하면 DB를 읽는 비용을 줄일 수 있습니다!

현재 구조상 DAO에게 전달하는 Reservation 객체는 생성자 필드로 ReservationTime과 Theme을 받게 되어있는데, 위와 같은 장점을 살리기 위해 날짜 검증을 Reservation 객체를 생성한 후에 수행하는 것보다는 RequestDTO를 받아온 시점에 제일 먼저 하면 어떨까 싶습니다. 😀

jbilee replied:
지난 날짜인지를 검증하려면 날짜뿐만 아니라, 요청으로 함께 들어온 timeId로 DB를 조회해 예약하려고 하는 시간대도 확인해야 한다는 점을 리팩터링을 하면서 깨달았습니다!

🤔 당일 예약은 허용하지 않는다는 정책을 가져가면 무조건 내일 날짜부터 예약이 가능하니 ReservationTime 없이 date만 비교할 수 있지만, 학습삼아 당일 예약도 가능하다는 가정 하에 리팩터링을 진행해봤습니다.

요청으로 들어온 timeId로 예약시간을 DB에서 받아온 다음 date와 time을 검증하고, 이후에 theme도 조회해서 Reservation 객체를 생성합니다. Time이 theme보다 먼저 필요해서 검증 로직이 중간에 들어가있지만, 따라가기 편한 플로우인지 궁금합니다!

[refactor: 예약 생성 로직 수정](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/8625021560d6377214d6c63594a011d20d6eb7d8)

solarsalt replied:
네 필요한 검증이니, 지금 로직도 좋아요 👍

solarsalt commented on src/main/java/roomescape/service/ReservationRequestFactory.java:
```diff
@@ -0,0 +1,52 @@
+import roomescape.repository.ThemeRepository;
+
+@Component
+public class ReservationRequestFactory {
```
줄리가 부여한 이 클래스의 역할은 무엇인가요?
이름으로는 ReservationRequest 을 생성하는 factory 로 보이는데,
제가 코드를 읽고 파악한 역할은 
Reservation DB에 저장하기 전에 모든 값을 검증하고 안전하게 저장하기 위한 준비를 하는 것으로 보였습니다.

리팩토링 가능 방향
- 역할에 맞는 이름을 부여해보면 좋겠어요
- 어떤 작업은 ReservationService에서하고, 어떤 작업은 이 클래스로 위임할지 분류해보면 더 좋겠습니다.

jbilee replied:
마지막에 Reservation 객체를 반환해준다는 결과만 보고 factory로 칭했는데, Reservation 객체를 생성하기 위해 붙는 여러 작업들이 실제로는 factory의 책임 범위를 넘어가게 만든다는 것을 깨달았습니다!

그렇다면 이렇게 개선하면 어떨까 싶습니다:
  - `ReservationRequestFactory`는 `BookService`라고 칭하고 `ReservationService`와 예약 관련 일을 분담하는 서비스로 만든다
  - `ReservationService`는 예약 목록 조회만 담당한다
  - `BookService`는 예약 저장과 삭제를 담당한다
  - `/reservations` 엔드포인트의 컨트롤러는 `ReservationService`와 `BookService`에 의존하도록 하고, 요청 HTTP 메서드에 따라 올바른 서비스에게 일을 시킨다

이렇게 구현하면 BookService가 지난 날짜 검증과 DB에서 시간&테마 조회 작업을 수행하기 적합한 객체가 될 것 같습니다!

jbilee replied:
[refactor: ReservationRequestFactory를 서비스 객체로 변경](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/809a1b2ed6c1ff98e150ee12291fb070c36e9328)

solarsalt replied:
ReservationService와 BookService 는 사실상 예약의 동의어라서 둘의 역할을 이름으로 구분하기 어려운 것 같아요. 

줄리가 정의한 역할을 유지하고 이름만 바꾼다면 저라면 아래 이름들이 더 어울리는 것 같고요. 
ReservationService는 예약 목록 조회만 담당한다 -> ReservationQueryService, ReservationFindService
BookService는 예약 저장과 삭제를 담당한다 -> BookUpdateService, BookCreateService

다른 역할을 설계한다면 저라면, 예약준비 -> 예약 이런식으로 구분해볼 거 같아용!
어려웠을텐데 서비스를 분리하는 시도를 해보신 것은 멋집니다!!👍

solarsalt commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -40,6 +41,12 @@ private void validateReservationTime(ReservationTime reservationTime) {
         }
     }
 
+    public boolean isBeforeCurrentDateTime() {
```
이 메소드를 외부에서 호출해주는 게 아니라, 생성 시점에 자체적으로 검증할 수도 있겠네요!
줄리 의견은 어떠신가요?

jbilee replied:
저도 동의합니다! 해서 Reservation 객체를 생성할 때 지난 날짜인지를 먼저 검증하는 로직을 적용해보기도 했는데요, 새로운 예약 데이터를 create할 땐 의도한대로 동작하지만 데이터베이스에 있는 예약을 read해올 때 read한 날보다 과거인 데이터면 DAO에서 Reservation 객체로 mapping할 때 예외가 발생해 클라이언트에 조회 실패로 응답이 가는 이슈가 있었습니다...😂 일반 사용자를 대상으로 한다면 항상 미래의 날짜로 예약된 데이터를 read해도 되겠지만, 어드민 페이지의 경우 과거 예약 내역까지 조회하는 기능이 필요할 것 같아서 다른 방법을 찾아야겠다고 생각했습니다!

처음엔 ReservationRequest라는 (도메인 내에서의) 리소스 생성 요청 전용 클래스를 별도로 만들어서 해당 인스턴스 생성 시에 지난 날짜인지를 검증하는 방안이 떠올랐는데, ReservationRequestFactory의 개선 방법을 고민하다가 Reservation 객체를 생성하기 전에 외부에서 지난 날짜인지를 검증하는 방향으로 수정해보면 좋겠다는 생각이 들었습니다. Reservation 객체를 생성하려면 ReservationTime과 Theme 객체까지 필요한데, 해당 객체들을 가져오려면 DB와의 통신이 필요해서요! DB를 실제로 조회하기 전에 지난 날짜인지를 먼저 확인해 early return하는 것이 더 효율적일 것 같습니다!

solarsalt replied:
여러 상황에서 이 객체의 검증이 필요하고, 안정적으로 성공하는지 고려하신 것 좋아요!
:) 생성시점 외에 다른 적합한 시점이나 객체로 이사를 고려해보시죠!

solarsalt commented on src/main/java/roomescape/controller/ThemeController.java:
```diff
@@ -0,0 +1,51 @@
+
+@RestController
+@RequestMapping("/themes")
+public class ThemeController {
```
테마 생성, 조회, 삭제 시에 기대하는 HTTP Status와 에러메시지가 오는지 검증하는 테스트 코드가 있다면 더 좋을 것 같아요!
현재 ThemeServiceTest 에서 로직은 검증할 수 있지만, 최종적으로 response가 기대하는 대로 동작하는지는 확인할 수가 없습니다~

jbilee replied:
새로 만든 ThemeController에 대한 테스트가 빠져있던 점 확인했습니다! 😅 피드백 감사합니다!

[test: ThemeController 테스트 추가](https://github.com/woowacourse/spring-roomescape-member/pull/244/commits/59459107cb911d5c778e5be474693662b8e347ae)

solarsalt reviewed (changes requested):
@jbilee 줄리~
1차로 작업해주신 내용은 거의 다 잘 반영되었어요! 
- 예약서비스를 리팩토링 해주신 새로운 클래스에서 어색한 부분에 코멘트 남겼습니다. 2차작업에서 개선해보시죠~
- 이번 미션의 핵심은 API 설계와 테스트에요~ 그래서 테마 API의 테스트가 보강된다면 좋겠습니다.
- 기대하는 에러메시지가 view에 보여지는 것은 우선순위를 낮춰서 선택적으로 작업해주셔도 됩니다.

jbilee committed:
refactor: URL 애노테이션에 protocol 속성 명시

jbilee committed:
chore: GlobalExceptionHandler 클래스 패키지 이동

jbilee committed:
refactor: 예외 전환 시 기존 예외를 cause로 전달하도록 수정

jbilee committed:
test: ReservationService 테스트에 누락된 예외 클래스 검증 추가

jbilee committed:
refactor: ReservationRequestFactory를 서비스 객체로 변경

- 클래스명을 BookService로 수정
- 예약 생성 및 삭제 메서드를 본 클래스로 이동

jbilee committed:
test: BookService 서비스 생성에 따른 테스트 코드 분리

jbilee committed:
refactor: IllegalStateException 사용 기준 변경

- IllegalStateException 예외는 인프라에서 리소스 저장이 실패할 때 현재 인프라가 정상동작하지 않는다는 의미로 사용
- 외래키로 인한 실패 케이스를 나타내기 위한 ConstrainedDataException 예외 클래스 생성
- IllegalStateException 발생 시의 상태코드를 503으로 변경

jbilee committed:
refactor: 예약 생성 로직 수정

- 시간과 테마를 DB에서 조회해서 Reservation 객체를 생성하는 대신, 시간을 먼저 조회하고 해당 시간과 요청값으로 지난 날짜인지를 서비스가 먼저 검증하도록 수정
- 동일한 예약 데이터가 존재하는지를 DAO에서 throw하는지 여부를 통해 확인하도록 수정

jbilee committed:
refactor: ExceptionHandler의 반환 객체 타입 일체화

- ResponseEntity 또는 ProblemDetail을 반환하던 부분을 ProblemDetail로 통일

jbilee committed:
test: 누락된 페이지 응답 테스트 추가

jbilee committed:
test: ThemeController 테스트 추가

jbilee committed:
test: Fake 객체 및 테스트 코드 보완

- 잘못된 입력값으로 인해 실패하던 테스트 수정
- Fake 객체가 올바른 예외 클래스를 반환하도록 수정

jbilee committed:
refactor: 뷰 코드 리팩터링

- 데이터 생성 또는 삭제가 실패할 경우 alert로 오류 메세지를 노출시키도록 수정

jbilee committed:
test: ThemeControllerTest 테스트 코드 수정

- MethodSource 값을 실제로 주입시키지 않았던 이슈 수정

solarsalt reviewed (approved):
줄리~
작업해 주신 내용은 코멘트로 댓글 남겼습니다 :)
4단계로 넘어가서 더 본격적으로 웹 앱 만들기를 경험해보시죠!
